ldms/src/ldmsd/ldmsd.h

```diff
@@ -755,57 +762,70 @@ struct avl_q_item {
 };
 TAILQ_HEAD(avl_q, avl_q_item);
 struct ldmsd_plugin {
-	char name[LDMSD_MAX_PLUGIN_NAME_LEN];
+	struct ldmsd_cfgobj cfgobj;
+	char name[LDMSD_MAX_PLUGIN_NAME_LEN]; /* plugin name (e.g. meminfo) */
 	struct avl_q avl_q;
 	struct avl_q kwl_q;
 	enum ldmsd_plugin_type {
 		LDMSD_PLUGIN_OTHER = 0,
 		LDMSD_PLUGIN_SAMPLER,
 		LDMSD_PLUGIN_STORE
 	} type;
-	struct ldmsd_plugin_cfg *pi;
+
+	int multi_instance:1; /* 0 if this is not a multi-instance plugin */
+	long reserve0:63;
+	long reserve1;
+
+	struct rbn __plugin1_rbn; /* For internal use */
+
+	char *libpath;
+
 	enum ldmsd_plugin_type (*get_type)(struct ldmsd_plugin *self);
 	int (*config)(struct ldmsd_plugin *self, struct attr_value_list *kwl, struct attr_value_list *avl);
 	void (*term)(struct ldmsd_plugin *self);
 	const char *(*usage)(struct ldmsd_plugin *self);
 };
+typedef struct ldmsd_plugin *ldmsd_plugin_t;
 
 struct ldmsd_sampler {
 	struct ldmsd_plugin base;
-	ldms_set_t (*get_set)(struct ldmsd_sampler *self);
-	int (*sample)(struct ldmsd_sampler *self);
-};
 
-struct ldmsd_plugin_cfg {
-	void *handle;
-	char *name;
-	char *libpath;
 	unsigned long sample_interval_us;
 	long sample_offset_us;
+
 	int thread_id;
-	int ref_count;
-	union {
-		struct ldmsd_plugin *plugin;
-		struct ldmsd_sampler *sampler;
-		struct ldmsd_store *store;
-	};
-	struct timeval timeout;
-	pthread_mutex_t lock;
 	ovis_scheduler_t os;
 	struct ovis_event_s oev;
-	LIST_ENTRY(ldmsd_plugin_cfg) entry;
+
+	ldms_set_t (*get_set)(struct ldmsd_sampler *self);
+	int (*sample)(struct ldmsd_sampler *self);
 };
-LIST_HEAD(plugin_list, ldmsd_plugin_cfg);
+#define LDMSD_SAMPLER(ptr) ((struct ldmsd_sampler *)(ptr))
+
 
 #define LDMSD_DEFAULT_SAMPLE_INTERVAL 1000000
 /** Metric name for component ids (u64). */
 #define LDMSD_COMPID "component_id"
 /** Metric name for job id number */
 #define LDMSD_JOBID "job_id"
 
-extern void ldmsd_config_cleanup(void);
-extern int ldmsd_config_init(char *name);
-struct ldmsd_plugin_cfg *ldmsd_get_plugin(char *name);
+struct ldmsd_plugin *ldmsd_get_plugin(const char *name);
+void ldmsd_put_plugin(struct ldmsd_plugin *pi);
```

Comment:
I am really not a fan of having both ldmsd_put_plugin and ldmsd_plugin_put. Too close in naming, and there is no documentation here about what they do either.